src/boot/efi/xbootldr.c

```diff
@@ -243,8 +234,15 @@ static EFI_STATUS find_device(EFI_HANDLE *device, EFI_DEVICE_PATH **ret_device_p
                 }
 
                 /* Patch in the data we found */
-                EFI_DEVICE_PATH *xboot_path = path_dup(partition_path);
-                memcpy((uint8_t *) xboot_path + ((uint8_t *) part_node - (uint8_t *) partition_path), &hd, sizeof(hd));
+                size_t len = (uint8_t *) part_node - (uint8_t *) partition_path;
+                EFI_DEVICE_PATH *xboot_path = xmalloc(len + sizeof(hd) + END_DEVICE_PATH_LENGTH);
+                memcpy(xboot_path, partition_path, len);
+
+                part_node = (EFI_DEVICE_PATH *) ((uint8_t *) xboot_path + len);
```

Comment:
btw, according to C, even such casts are illegal if the pointer is not aligned. Because the compiler automatically assumes alignment once a pointer has a certain type, and will generate optimized code from it. It's not legal to have an unaligned pointer of a specific type even if you never actually access it.

Reply:
I know. `EFI_DEVICE_PATH` has alignment of 1 though. So it's fine.

Reply:
this deserves a comment. Or an assertion:

```c
assert_cc(alignof(EFI_DEVICE_PATH) == 1);
```